editoast: remove 'validator' dependency

bougue-pe opened the pull request:
Supersedes most part of https://github.com/OpenRailAssociation/osrd/pull/10273
(Except the model change required as tracked in https://github.com/OpenRailAssociation/osrd/issues/13333)

* Remove 'locked' from RollingStockForm and TowedRollingStockForm in API 
* Few minor front improvement (first 2 commits + last 1)
* Use `editoast_schemas::RollingStock` instead of `RollingStockForm` (same for TowedRS), then validate during deserialize
* Add a 'locked' query-param to TowedRollingStock's creation (POST), to make it consistent with RollingStock
* Update front and test-fixtures accordingly
* Update and bump osrd-schemas version to `0.9.0`

✅ Hand-tested:
- manually create RS in front (duplicate, use in scenario, modify, delete)
- import a RS using editoast CLI, then modify it (even for an ETCS rolling-stock)
- manually change "locked" value in DB and check the front (forbidden edition/delete)
- export > import > export rolling-stock in front (except for a [previously existing bug](https://github.com/OpenRailAssociation/osrd/issues/13088))

🔍 Advise review by commit, and "hide whitespaces" for `import_rolling_stock.rs`.

TODO:
- [x] check attributes moving in openapi (`locked `+ `railjson_version`)
- [x] fix front (RS creation leads to "no `locked` field provided")*
- [x] adjust/test front for the possible other issues (unknown so far)
- [x] maybe do the same for `TowedRollingStock -> Done
- [x] check if osrd-data must be updated after these changes -> change patch to put is necessary and `locked` removal too, but no other change as `railjson_version` is defaulted.
- [x] merge https://github.com/OpenRailAssociation/osrd/pull/13086
- [x] be ready to merge on client's repository

bougue-pe committed:
editoast: remove validator crate and use schemas::RollingStock

* Use editoast_schemas::RollingStock instead of RollingStockForm
* Validate during deserialize
* Update front accordingly

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe committed:
front: add * to tell primary category is mandatory for RS

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe committed:
ci, editoast: update interface-change check after renames

after https://github.com/OpenRailAssociation/osrd/pull/12847

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

woshilapin commented on editoast/openapi.yaml:
```diff
@@ -11817,6 +11817,7 @@ components:
       type: object
       required:
       - name
+      - locked
```
It does require the `locked` to be present in the payload now, it wasn't mandatory before, so it is a breaking change. How about making that field removed first, before merging this PR?

bougue-pe replied:
After discussion, I made it optional (default to false): https://github.com/OpenRailAssociation/osrd/pull/12904/commits/c1c7ba40dc38849adc2d9c1de59d290a9be0e516#diff-dae42f0fbf235f43a761b4646e9701d4bb7c9d4409562e2aec2a1ee7fc251d4dR75-R79
This way, no breaking change is done to the API currently.

woshilapin commented on editoast/schemas/src/rolling_stock.rs:
```diff
@@ -57,15 +57,24 @@ use common::units::quantities::Velocity;
+#[editoast_derive::annotate_units]
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, ToSchema)]
 #[serde(remote = "Self")]
+#[schema(as = RollingStockForm)]
```
Shouldn't we remove that line? It might change more code, but I don't understand why we would need it?

bougue-pe replied:
This must match what's used in `views/rolling_stock.rs` (probably annotations).

Resolving this comment, and discussion may continue in the ["views" comment](https://github.com/OpenRailAssociation/osrd/pull/12904#discussion_r2315925303).

woshilapin commented on editoast/src/views/rolling_stock.rs:
```diff
@@ -1,8 +1,7 @@
 pub(in crate::views) mod towed;
 
-pub use form::RollingStockForm;
+pub use schemas::RollingStock as RollingStockForm;
```
If you remove the `schema` rename for OpenAPI, you also might want to remove it here. If not, feel free to resolve either way.

bougue-pe replied:
As discussed, there are 2 `RollingStock` exposed in the openAPI (`models` as output, and `schemas` now as input).

So I tried just using it on all inputs through the namespace, completely removing this `pub use`, and annotating `schemas::RollingStock` (see this [comment](https://github.com/OpenRailAssociation/osrd/pull/12904#discussion_r2315917006)) with `#[schema(as = schemas::RollingStock)]`.
It does compile, tests OK and correctly generates openapi and derivatives (didn't dive to change all the front accordingly and test).

The conterpart is that the changes are much less readable as `RollingStockForm` is the old name, and `schema.RollingStock` is a bit weird reference in openapi (correctly derived to `SchemasRollingStock` in ts).

I'd vote (lazily) to keep it as-is, but it seems like doing it in a separate commit would make it clear enough to review: (anyone with an opinion) let me know if you'd rather keep it as I vote (👍) or change it (👎).

---

On a separate note, it underlines that `pub` is probably unnecessary, will remove it: done in https://github.com/OpenRailAssociation/osrd/pull/12904/commits/c1c7ba40dc38849adc2d9c1de59d290a9be0e516#diff-7b687915e99ad9379bf3c3f132b4ffdb88443d7680d3b4a46c3cb2061c7e678eR4

woshilapin replied:
Let's keep it as is.

woshilapin reviewed (commented):
Just love the branch's name!

I'm wondering if removing the `locked` field from schema wouldn't be a better idea to change the API in a compatible way?

Math-R commented on front/src/applications/rollingStockEditor/components/RollingStockEditorForm.tsx:
```diff
@@ -82,7 +81,7 @@ const RollingStockEditorForm = ({
   const addNewRollingstock = (payload: RollingStockForm) => () => {
     postRollingstock({
       locked: false,
-      rollingStockForm: payload,
+      rollingStockForm: { ...payload, locked: false },
```
It seems redundant to pass ` locked: false` both at the root of the payload and within rollingStockForm.

Is there a specific reason for keeping both? Are they meant to serve different purposes at each level of the payload structure?

If not, could we simplify by removing one of them, maybe the one at the root to avoid confusion?

woshilapin replied:
One will be removed, hence my remark in https://github.com/OpenRailAssociation/osrd/pull/12904#discussion_r2315912435. But the PR is already big so I believe @bougue-pe planned to not do it in the current PR.

bougue-pe replied:
To clarify a bit: this duplicate was already there (only less explicit).
As [mentionned](https://github.com/OpenRailAssociation/osrd/pull/12904#discussion_r2319398532), I made it optional again, so I removed pretty much all the changes to the front (except that it's an official PUT now, not a PATCH anymore).

So the current PR doesn't address all, but it doesn't worsen anything, and only improves the crate part: it can be merged without addressing duplicate IMO.

The idea is to do a subsequent PR to completely remove the attribute from the form (and update clients of the API).

Math-R replied:
Okay, fine :)

hamz2a reviewed (approved):
LGTM, thanks!

Khoyo reviewed (approved):
CI LGTM

flomonster reviewed (commented):
LGTM except the added field `locked` to the form.

Math-R reviewed (approved):
lgtm

bougue-pe committed:
editoast: remove 'locked' from RollingStockForm in API

'locked' is actually removed from editoast-schemas

* bump osrd-schemas version to  0.9.0
* update test fixtures
* modify osrd-schemas accordingly

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe committed:
editoast: remove 'locked' from TowedRollingStockForm in API

Actually start using schemas::TowedRollingStock in place of
TowedRollingStockForm ('locked' is removed from editoast-schemas).

Add a 'locked' query-param to TowedRollingStock's creation (POST), to make
it consistent with RollingStock's creation (used the same way by clients).

* update test fixtures
* modify osrd-schemas accordingly
* after previous commit, additional version bumps of osrd-schemas are not
  necessary.

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe committed:
front: rename translation key

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>

bougue-pe commented:
Just some minor changes where just added after last reviews (that included `fixup!` commit): https://github.com/OpenRailAssociation/osrd/compare/c1c7ba40dc38849adc2d9c1de59d290a9be0e516..b9fbc4536bee5dfafff3e0821360ade55ec2ea53

A PR on top of this one is going to be merged before merging all to dev: https://github.com/OpenRailAssociation/osrd/pull/13086.

And a MR on editoast's client is also ready.

The reason to group it is that it's a breaking change anyway to change the verb, so let's avoid multiple bumps especially on the client's side.

flomonster commented:
Can we rebase this PR and merge it?
